### PR TITLE
Architecture Documentation

### PR DESCRIPTION
**Pull request by jsingle**
@thieman I'm probably missing stuff that should be discussed.  Also I can make a diagram, that might be nice

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Dusty Architecture
+
+## Client / Daemon Structure
+
+Dusty uses a client/daemon structure, where a daemon (`dusty -d`) runs with root permissions,
+and a client sends it commands over a unix socket. By default the daemon is run by a plist
```
Comment (thieman) [inline docs/architecture.md]: Runs as root? Why the Hell does it have to do that?

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+`/System/Library/LaunchDaemons/org.gamechanger.dusty.plist`; the client is just run on the
```
Comment (thieman) [inline docs/architecture.md]: Too detailed maybe?

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+## System Components
+
+Dusty leverages several programs and system components:
+
+ * your hosts file
+ * nginx
+ * docker
+ * boot2docker
+ * docker-compose
```
Comment (thieman) [inline docs/architecture.md]: Probably worth mentioning VirtualBox

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+### Overview
+
+Dusty will use docker-compose to stitch together and launch your apps
+and services.  These run in docker containers inside your boot2docker VM.  Nginx allows
+you to talk to these apps, by routing your local requests to intermediate ports on the
+boot2docker VM. These map to any port on your app's docker container that your app would
+like to use.
```
Comment (thieman) [inline docs/architecture.md]: This section would definitely be aided by a diagram, yeah.

File: docs/architecture.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+# BEGIN section for Dusty
+127.0.0.1 local.example-app.com
+# END section for Dusty
+```
+This allows you to visit `local.example-app.com` in your browser, `curl` it, etc.
+
+### Nginx
+
+Dusty writes a `dusty.conf` nginx configuration file, which is placed in a directory
+that nginx includes (eg `/usr/local/etc/nginx/servers/`).
```
Comment (thieman) [inline docs/architecture.md]: e.g.

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+### Hosts File
+
+Your hosts file (`/etc/hosts`) is modified by Dusty so that you can use Dusty-specified
+local host names.  An example of the additions that Dusty makes to your hostfile is:
+```
+# BEGIN section for Dusty
+127.0.0.1 local.example-app.com
+# END section for Dusty
+```
+This allows you to visit `local.example-app.com` in your browser, `curl` it, etc.
```
Comment (thieman) [inline docs/architecture.md]: Well, this is part of what allows you to visit the URL.

File: docs/architecture.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+### Nginx
+
+Dusty writes a `dusty.conf` nginx configuration file, which is placed in a directory
+that nginx includes (eg `/usr/local/etc/nginx/servers/`).
+
+Nginx is used to route requests on your localhost to your boot2docker VM. Virtualbox
+port forwarding is not used, instead the actual ip of the VM is used directly.  Ports
```
Comment (thieman) [inline docs/architecture.md]: IP

File: docs/architecture.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+are mapped from an arbitrary port on your localhost (specified in your Dusty specs) to
```
Comment (thieman) [inline docs/architecture.md]: If it's specified then it's not arbitrary :ghost:

File: docs/architecture.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+         }
+     }
+}
+```
+
+### Boot2docker Virtual Machine
+
+#### Rsync
+
+Docker is run on OSX via the boot2docker virtual machine (managed by virtualbox).  So docker containers are actually running on the linux virtual machine.  These containers, running Dusty apps and services, need to access code living on your mac.  One way to do this is to use a
```
Comment (thieman) [inline docs/architecture.md]: Is RTD gonna do wrapping for us? We should probably pick a style and stick with it, this is the only line that isn't wrapped.

Comment (thieman) [inline docs/architecture.md]: > Docker is run on OSX via the boot2docker virtual machine (managed by virtualbox)

This is hard to understand, Docker (daemon) isn't really running *on* OS X at all. But Docker (client) does. How can we explain this better? Also this seems like it shouldn't be in the rsync section.

Comment (jsingle) [inline docs/architecture.md]: RTD does line wrap, but this was a mistake.  I prefer to not have giant lines like this in source

File: mkdocs.yml [hunk 10]
```diff
@@ -7,3 +7,4 @@ pages:
   - Home: index.md
   - Changelog: changelog.md
   - Contributing: contributing.md
+  - Architecture: architecture.md
```
Comment (thieman) [inline mkdocs.yml]: Scootch this on up under Home.

File: docs/architecture.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+Virtualbox shared folder; however the performance of shared folders is very poor.  As an
+alternative, we use rsync to move files from your mac to the VM.
```
Comment (thieman) [inline docs/architecture.md]: We simply must link to that awesome benchmark blog post to show people just how slow shared folders are.

File: docs/architecture.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+Dusty stores persistent data in the folder `/persist` on the VM, which is symlinked to
+boot2docker's persistent virtual disk.
+
+### Docker-compose
+
+[Docker-compose](https://docs.docker.com/compose/) is a tool for defining and runnning
+multi-container applications with docker.  It's used to specify docker entrypoints, volumes,
+linked containers, and more.  By using docker-compose linked containers, Dusty apps can
+talk to other apps and services simply by using the names of those apps and services as
+hostnames.
```
Comment (thieman) [inline docs/architecture.md]: Should talk about our big value add on Docker Compose here: our DAG and how we stitch together composefiles.

**Pull request merged**